File: backend/internal/httpapi/images.go
package httpapi

import (
	"encoding/base64"
	"fmt"

	"github.com/abubakarsiddik31/golem/model"
)

const (
	maxImagesPerMessage = 4
	maxImageBytes       = 4 << 20 // per decoded image
)

var allowedImageTypes = map[string]bool{
	"image/png":  true,
	"image/jpeg": true,
	"image/webp": true,
	"image/gif":  true,
}

// imageInput is one base64-encoded attachment in a send-message body.
type imageInput struct {
	MediaType string `json:"mediaType"`
	Data      string `json:"data"` // standard base64, no data-URL prefix
}

// decodeImages validates and decodes attachments into golem image parts.
// Inline bytes (not URLs) are the portable path: Gemini's URL form expects
// provider-reachable file URIs, while inline base64 works on every provider.
func decodeImages(in []imageInput) ([]model.Part, error) {
	if len(in) == 0 {
		return nil, nil
	}
	if len(in) > maxImagesPerMessage {
		return nil, fmt.Errorf("at most %d images per message", maxImagesPerMessage)
	}
	parts := make([]model.Part, 0, len(in))
	for i, img := range in {
		if !allowedImageTypes[img.MediaType] {
			return nil, fmt.Errorf("image %d: unsupported type %q", i+1, img.MediaType)
		}
		data, err := base64.StdEncoding.DecodeString(img.Data)
		if err != nil {
			return nil, fmt.Errorf("image %d: invalid base64", i+1)
		}
		if len(data) == 0 {
			return nil, fmt.Errorf("image %d: empty", i+1)
		}
		if len(data) > maxImageBytes {
			return nil, fmt.Errorf("image %d: larger than %d bytes", i+1, maxImageBytes)
		}
		parts = append(parts, model.ImageData(img.MediaType, data))
	}
	return parts, nil
}

File: frontend/src/lib/images.ts
// Client-side mirror of the backend's image attachment limits (see
// backend/internal/httpapi/images.go) — reject early, before upload.

export const MAX_IMAGES_PER_MESSAGE = 4;
export const MAX_IMAGE_BYTES = 4 << 20; // per decoded image

export const ALLOWED_IMAGE_TYPES = ["image/png", "image/jpeg", "image/webp", "image/gif"] as const;

export interface PendingImage {
  mediaType: string;
  /** Standard base64, no data-URL prefix — the wire format. */
  data: string;
  /** Object URL for the composer preview. */
  previewUrl: string;
  name: string;
}

// readImageFiles validates and decodes picked image files. It stops at the
// per-message cap and reports every problem file; valid files still load.
export async function readImageFiles(files: File[]): Promise<{ images: PendingImage[]; errors: string[] }> {
  const images: PendingImage[] = [];
  const errors: string[] = [];
  for (const file of files) {
    if (images.length >= MAX_IMAGES_PER_MESSAGE) {
      errors.push(`At most ${MAX_IMAGES_PER_MESSAGE} images per message.`);
      break;
    }
    if (!(ALLOWED_IMAGE_TYPES as readonly string[]).includes(file.type)) {
      errors.push(`${file.name}: unsupported type.`);
      continue;
    }
    if (file.size > MAX_IMAGE_BYTES) {
      errors.push(`${file.name}: larger than 4 MB.`);
      continue;
    }
    try {
      const dataUrl = await readFileAsDataURL(file);
      images.push({
        mediaType: file.type,
        data: dataUrl.slice(dataUrl.indexOf(",") + 1),
        previewUrl: URL.createObjectURL(file),
        name: file.name,
      });
    } catch {
      errors.push(`${file.name}: could not be read.`);
    }
  }
  return { images, errors };
}

function readFileAsDataURL(file: File): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = () => resolve(String(reader.result));
    reader.onerror = () => reject(reader.error);
    reader.readAsDataURL(file);
  });
}
